Model plugin emits unused uuid import for tables without UUID columns

# Model plugin emits unused `uuid` import for tables without UUID columns

## Summary

When generating models with `bobgen-psql`, the model plugin adds an
`import "github.com/google/uuid"` to the generated `*.bob.go` file for tables
that have **no UUID columns**. The import is never referenced in the file, so
the generated package fails to compile with `imported and not used`.

## Version

- Generator: `BobGen psql v0.48.0` (header: `// Code generated by BobGen psql v0.48.0. DO NOT EDIT.`)
- Bob: `github.com/stephenafamo/bob v0.48.0`
- UUID opt lib: `github.com/aarondl/opt v0.0.0-20250607033636-982744e1bd65`
- Go: `go1.26`
- OS: macOS

## Configuration (bobgen.yaml)

```yaml
psql:
  dialect: psql
  driver: github.com/jackc/pgx/v5/stdlib
  schemas: ["public", "app"]
  shared_schema: "app"
  pattern: "files/schema.sql"
  uuid_pkg: google
  # ...
```

All model/factory/where/loaders/joins plugins are enabled (`plugins_preset: all`).

## Affected table

The table has only scalar columns and no UUID anywhere:

```sql
CREATE TABLE sample_table (
    id         integer PRIMARY KEY,
    parent_id  integer NOT NULL,
    label_id   integer NOT NULL,
    enabled    boolean NOT NULL
);
```

## Generated output (excerpt)

```go
// Code generated by BobGen psql v0.48.0. DO NOT EDIT.

package models

import (
	"context"
	"fmt"
	"io"
	"strconv"

	"github.com/aarondl/opt/omit"
	"github.com/google/uuid"      // <-- added but never used
	"github.com/stephenafamo/bob"
	// ...
)
```

`grep` confirms `uuid.` is referenced 0 times in the generated file, while it is
correctly imported and used in models for tables that actually have UUID columns.

## Actual behavior

`go build ./db/models/` fails:

```
db/models/sample_table.bob.go:13:2: "github.com/google/uuid" imported and not used
```

## Expected behavior

The `uuid` import should only be emitted when the generated file actually
references the uuid package (i.e. the table has a UUID column, or the factory /
random-value code for that table needs it). Tables without UUID columns should
not import `github.com/google/uuid`.

## Reproduction steps

1. Define a table with only non-UUID columns (see `sample_table` above).
2. Set `uuid_pkg: google` in the config.
3. Run `bobgen-psql -c bobgen.yaml`.
4. `go build ./...` on the generated models package.

## Notes / suspected cause

The stray import appears to come from the import set of the factory /
random-value template being merged into the model file's import block
unconditionally, rather than being gated on whether the table actually uses the
uuid type. (Only the `uuid` import is affected; other conditional imports like
`strconv`/`io` are correctly present because they are used by the random-value
helpers.)

## Current workaround

Post-process generated output with `goimports -w` (gofumpt alone does not remove
unused imports) or `golangci-lint run --fix` to strip the unused import after
generation.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -20,6 +20,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Fixed
 
+- Fixed an unused type import (e.g. `github.com/google/uuid`) in generated models when the type is only referenced through optional wrapper expressions, such as m2m relationship helpers for a join table with a `uuid` column ([#730](https://github.com/stephenafamo/bob/issues/730)).
 - Fixed the counts plugin's generated `C` field reusing the `relation_tag` struct tag, which collided with the relations `R` field and caused `encoding/json` to drop both fields when a real tag name was configured. The `C` field now uses a dedicated `relation_tag_count` config option (default `"-"`). (thanks @jacobmolby)
 - Fixed generated join alias suffixes using `randInt()`, which could produce negative values when `maphash.Hash.Sum64()` overflowed `int64` ([#719](https://github.com/stephenafamo/bob/issues/719)). Generated join mods now use `bob.NextUniqueInt()` instead. (thanks @atzedus)
 - Fixed `orm.Query.Clone()` dropping the `Scanner`, which made `All`/`One`/`Cursor`/`Each` on a cloned query panic with a nil pointer dereference. (thanks @sandonemaki)

--- a/gen/bobgen-psql/driver/gen_unused_import_test.go
+++ b/gen/bobgen-psql/driver/gen_unused_import_test.go
@@ -1,0 +1,146 @@
+package driver_test
+
+import (
+	"context"
+	"os"
+	"path/filepath"
+	"strings"
+	"testing"
+
+	"github.com/stephenafamo/bob/gen"
+	helpers "github.com/stephenafamo/bob/gen/bobgen-helpers"
+	driver "github.com/stephenafamo/bob/gen/bobgen-psql/driver"
+	"github.com/stephenafamo/bob/gen/drivers"
+	"github.com/stephenafamo/bob/gen/plugins"
+)
+
+type stubDriver struct {
+	types drivers.Types
+	info  *drivers.DBInfo[any, any, driver.IndexExtra]
+}
+
+func (s stubDriver) Dialect() string      { return "psql" }
+func (s stubDriver) Types() drivers.Types { return s.types }
+func (s stubDriver) Assemble(context.Context) (*drivers.DBInfo[any, any, driver.IndexExtra], error) {
+	return s.info, nil
+}
+
+func col(name, dbType, goType string) drivers.Column {
+	return drivers.Column{Name: name, DBType: dbType, Type: goType}
+}
+
+// Mirrors issue #730: a table with no UUID columns must not import the uuid
+// package in its generated model file, even when uuid_pkg is configured and
+// other tables use UUIDs.
+func TestScalarTableModelHasNoUUIDImport(t *testing.T) {
+	types := helpers.Types()
+	types.Register("uuid.UUID", drivers.Type{
+		Imports:    []string{`"github.com/google/uuid"`},
+		RandomExpr: `return uuid.New()`,
+	})
+
+	info := &drivers.DBInfo[any, any, driver.IndexExtra]{
+		Driver: "github.com/jackc/pgx/v5/stdlib",
+		Tables: drivers.Tables[any, driver.IndexExtra]{
+			{
+				Key:  "sample_table",
+				Name: "sample_table",
+				Columns: []drivers.Column{
+					col("id", "integer", "int32"),
+					col("parent_id", "integer", "int32"),
+					col("label_id", "integer", "int32"),
+					col("enabled", "boolean", "bool"),
+				},
+				Constraints: drivers.Constraints[any]{
+					Primary: &drivers.Constraint[any]{Name: "sample_table_pkey", Columns: []string{"id"}},
+				},
+			},
+			{
+				Key:  "tags",
+				Name: "tags",
+				Columns: []drivers.Column{
+					col("id", "uuid", "uuid.UUID"),
+					col("name", "text", "string"),
+				},
+				Constraints: drivers.Constraints[any]{
+					Primary: &drivers.Constraint[any]{Name: "tags_pkey", Columns: []string{"id"}},
+				},
+			},
+			{
+				Key:  "sample_tags",
+				Name: "sample_tags",
+				Columns: []drivers.Column{
+					col("sample_id", "integer", "int32"),
+					col("tag_id", "uuid", "uuid.UUID"),
+				},
+				Constraints: drivers.Constraints[any]{
+					Primary: &drivers.Constraint[any]{Name: "sample_tags_pkey", Columns: []string{"sample_id", "tag_id"}},
+					Foreign: []drivers.ForeignKey[any]{
+						{
+							Constraint:     drivers.Constraint[any]{Name: "sample_tags_sample_id_fkey", Columns: []string{"sample_id"}},
+							ForeignTable:   "sample_table",
+							ForeignColumns: []string{"id"},
+						},
+						{
+							Constraint:     drivers.Constraint[any]{Name: "sample_tags_tag_id_fkey", Columns: []string{"tag_id"}},
+							ForeignTable:   "tags",
+							ForeignColumns: []string{"id"},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	oldWD, err := os.Getwd()
+	if err != nil {
+		t.Fatal(err)
+	}
+	tmp := t.TempDir()
+	gomod := "module scratch.local/repro\n\ngo 1.24\n"
+	if err := os.WriteFile(filepath.Join(tmp, "go.mod"), []byte(gomod), 0o600); err != nil {
+		t.Fatal(err)
+	}
+	if err := os.Chdir(tmp); err != nil {
+		t.Fatal(err)
+	}
+	t.Cleanup(func() { _ = os.Chdir(oldWD) })
+
+	outputPlugins := plugins.Setup[any, any, driver.IndexExtra](
+		plugins.PresetAll, gen.PSQLTemplates,
+	)
+
+	state := &gen.State[any]{Config: gen.Config[any]{}}
+	if err := gen.Run[any, any, driver.IndexExtra](
+		context.Background(), state, stubDriver{types: types, info: info}, outputPlugins...,
+	); err != nil {
+		t.Fatalf("gen.Run: %v", err)
+	}
+
+	var checked int
+	err = filepath.WalkDir(tmp, func(path string, d os.DirEntry, err error) error {
+		if err != nil || d.IsDir() || !strings.HasSuffix(path, ".go") {
+			return err
+		}
+		b, err := os.ReadFile(path)
+		if err != nil {
+			return err
+		}
+		rel, _ := filepath.Rel(tmp, path)
+		if strings.Contains(rel, "sample_table") {
+			checked++
+			if strings.Contains(string(b), "github.com/google/uuid") &&
+				!strings.Contains(string(b), "uuid.") {
+				t.Errorf("%s imports github.com/google/uuid but never uses it", rel)
+			}
+			t.Logf("checked %s (%d bytes)", rel, len(b))
+		}
+		return nil
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+	if checked == 0 {
+		t.Fatal("no sample_table files were generated/checked")
+	}
+}

--- a/gen/drivers/types.go
+++ b/gen/drivers/types.go
@@ -351,7 +351,9 @@ func (t Types) FromOptional(currentPkg string, i language.Importer, forType, var
 	colTyp, _ := t.GetNameAndDef(currentPkg, forType)
 	optTyp, imports := t.getOptional(currentPkg, forType, isNull, fromOrToNull)
 	nullTyp := t.GetNullType(currentPkg, forType)
-	i.ImportList(imports)
+	if exprNamesType(optTyp.UseExpr) {
+		i.ImportList(imports)
+	}
 	i.ImportList(optTyp.UseExprImports)
 	return strings.NewReplacer(
 		"SRC", varName,
@@ -365,7 +367,9 @@ func (t Types) ToOptional(currentPkg string, i language.Importer, forType, varNa
 	colTyp, _ := t.GetNameAndDef(currentPkg, forType)
 	optTyp, imports := t.getOptional(currentPkg, forType, isNull, fromOrToNull)
 	nullTyp := t.GetNullType(currentPkg, forType)
-	i.ImportList(imports)
+	if exprNamesType(optTyp.CreateExpr) {
+		i.ImportList(imports)
+	}
 	i.ImportList(optTyp.CreateExprImports)
 	return strings.NewReplacer(
 		"SRC", varName,
@@ -373,6 +377,17 @@ func (t Types) ToOptional(currentPkg string, i language.Importer, forType, varNa
 		"BASETYPE", colTyp,
 		"OPTIONALTYPE", optTyp.Name,
 	).Replace(optTyp.CreateExpr)
+}
+
+// exprNamesType reports whether an optional/null expression template writes a
+// type name into the generated code. Expressions like `omit.From(SRC)` only
+// reference the wrapped value, so the base type's imports are not needed for
+// them; expressions like `func () *BASETYPE {...}` or `NULLTYPE{V: SRC}` spell
+// out a type name and do need them.
+func exprNamesType(expr string) bool {
+	return strings.Contains(expr, "BASETYPE") ||
+		strings.Contains(expr, "NULLTYPE") ||
+		strings.Contains(expr, "OPTIONALTYPE")
 }
 
 var _ TypeModifier = AarondlNull{}

